sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala

```diff
@@ -361,7 +361,7 @@ case class StreamingDeduplicateExec(
       val numUpdatedStateRows = longMetric("numUpdatedStateRows")
 
       val baseIterator = watermarkPredicate match {
-        case Some(predicate) => iter.filter((row: InternalRow) => !predicate.eval(row))
+        case Some(predicate) => iter.filter(row => !predicate.eval(getKey(row)))
```

Comment:
This is very minor, but maybe we should just bind a different predicate to the `child.output` schema rather than have two projections here.

Perhaps the base class should have both `watermarkExpression` and `keyPredicate` to allow for this flexibility.